docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+* Prefer single (lowercase) words for package names where feasible
+* Use __lowercase with underscore__ (a.k.a "snake_case") to split words in package names if needed
+
+I believe this would result in Crate names (which must be Rust identifiers) being identical to the package name at all times.
+
+## Module Layout
+
+Modules form the basic units of imported code in Rust, so identifying where to split functionality and what to call modules is essential. I suspect this is the right level to divide things like collation vs segmentation etc. but it's also far too early to make hard recommendations about this, so I proposed to leave this issue open for the time being (moving code between modules later should be relatively easy).
+
+**Open Question**: Should we have code in the top-level module of a Crate or always use named sub-modules? This probably ties into what Crates we intend to publish and the granularity of the APIs.
+
+## Shared Internal Code
+
+Rust supports having non-public sub-modules within a Crate, and only the top-level "implicit" module is exposed by default (other modules can be exported by the top level module). See also [File hierarchy - Rust By Example](https://doc.rust-lang.org/rust-by-example/mod/split.html).
+
+### Use Private Modules :: required
+
+While local private functions can just be defined in any module, shared internal functions should always be implemented in a non-exported sub-module. This avoids any misleading dependencies between "public" modules and should help ensure clean code dependencies. Code in internal modules should still be tested and documented thoroughly. It is perfectly expected that we would have multiple internal modules for different aspects of ICU4X.
+
+# Structs and Traits
+
+For ICU4X we should be looking to keep code consistent and very unsurprising, especially for layout and behavior of data structures.
+
+## Private vs Public
+
+Rust offers a compelling model for encapsulating implementation details based on a hierarchical view of privacy; see [Visibility and Privacy - The Rust Reference](https://doc.rust-lang.org/reference/visibility-and-privacy.html). This should allow ICU4X to encapsulate all of its implementation details neatly from applications using the library and avoid too many cases of [Hyrum's Law](https://www.hyrumslaw.com/).
+
+### Minimize user-visible structs :: required
+
+While this sounds a bit obvious, I think it's important to stress that with Rust's privacy model, there is never any need to have an internal type be user-visible, and once a type is user-visible outside a Crate, it's published and should not be changed.
+
+The only place this might be an issue is if we decide to deliver ICU4X as multiple Crates. If we do that then we just added a requirement that our inter-Crate APIs be public (there's no privileged sharing outside Crates). This is one reason I propose we deliver ICU4X as a single Crate with multiple modules.
```

Comment:
We can use `doc(hidden)` and modules starting with multiple underscores for this. This is a known rust pattern and people won't use APIs like this. It's often necessary to have public-hidden modules to make macros work, for example.

I would not choose the single-crate model just because of this.

Reply:
Done.